edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java

```diff
@@ -3269,5 +3290,93 @@ public Object doInHibernate(Session session) throws HibernateException {
 		return rval;
 		
 	}
-	
+
+
+	/**
+	 * Updates the categorized order of an assignment
+	 *
+	 * @see GradebookService.updateAssignmentCategorizedOrder(java.lang.String gradebookUid, java.lang.Long assignmentId, java.lang.Integer order)
+	 */
+	@Override
+	public void updateAssignmentCategorizedOrder(final String gradebookUid, final Long categoryId, final Long assignmentId, Integer order) {
+
+		if (!getAuthz().isUserAbleToEditAssessments(gradebookUid)) {
+			log.error("AUTHORIZATION FAILURE: User " + getUserUid() + " in gradebook " + gradebookUid + " attempted to change the order of assignment " + assignmentId);
+			throw new SecurityException("You do not have permission to perform this operation");
+		}
+
+		if(order == null) {
+			throw new IllegalArgumentException("Categoryized Order cannot be null");
+		}
+
+		final Long gradebookId = getGradebook(gradebookUid).getId();
+
+		//get all assignments for this gradebook
+		List<Assignment> assignments = getAssignments(gradebookId, SortType.SORT_BY_SORTING, true);
+		List<Assignment> assignmentsInNewCategory = new ArrayList<Assignment>();
+		for (Assignment assignment : assignments) {
+			if (assignment.getCategory() == null) {
+				if (categoryId == null) {
+					assignmentsInNewCategory.add(assignment);
+				}
+			} else if (assignment.getCategory().getId().equals(categoryId)) {
+				assignmentsInNewCategory.add(assignment);
+			}
+		}
+
+		//adjust order to be within bounds
+		if(order < 0) {
+			order = 0;
+		} else if (order > assignmentsInNewCategory.size()) {
+			order = assignmentsInNewCategory.size();
+		}
+
+		//find the assignment
+		Assignment target = null;
+		for(Assignment a: assignmentsInNewCategory){
+			if(a.getId().equals(assignmentId)) {
+				target = a;
+				break;
+			}
+		}
+
+		//add the assignment to the list via a 'pad, remove, add' approach
+		assignmentsInNewCategory.add(null); //ensure size remains the same for the remove
+		assignmentsInNewCategory.remove(target); //remove item
+		assignmentsInNewCategory.add(order, target); //add at ordered position, will shuffle others along
+
```

Comment:
Good work reusing this stuff from the GradebookNgBusinessService. Took me a while to figure out this shuffle approach, had it all sketched out on paper, ha!